design/UsageValidationRules.md

```diff
@@ -0,0 +1,71 @@
+## Definition of "resources being used"
+
+Resources in WebGPU can be used in different ways that are declared at resource creation time and interact with various validation rules.
+
+Buffers have the following usages (and commands inducing that usage):
+ - `VERTEX` for the `buffers` in calls to `WebGPUCommandEncoder.setVertexBuffers`
+ - `INDEX` for `buffer` in calls to `WebGPUCommandEncoder.setIndexBuffer`
+ - `INDIRECT` for `indirectBuffer` in calls to `WebGPUCommandEncode.{draw|drawIndexed|dispatch}Indirect`
+ - `UBO`, `RO_STORAGE` and `STORAGE` for buffers referenced by bindgroups passed to `setBindGroup`, with the usage corresponding to the binding's type.
+ - `TRANSFER_SRC` for buffers used as the copy source of various commands.
+ - `TRANSFER_DST` for buffers used as the copy destination of various commands.
+ - (Maybe [RO_]STORAGE_TEXEL and UNIFORM_TEXEL?)
+
+Textures have the following usages:
+ - `OUTPUT_ATTACHMENT` for the subresources referenced by `WebGPURenderPassDescriptor`
+ - `SAMPLED`, `RO_STORAGE` and `STORAGE` for subresources corresponding to the image views referenced by bindgroups passed to `setBindGroup`, with the usage corresponding to the binding's type.
+ - `TRANSFER_SRC` for textures used as the copy source of various commands.
+ - `TRANSFER_DST` for textures used as the copy destination of various commands.
+ 
+Read only usages are `VERTEX`, `INDEX`, `INDIRECT`, `UBO`, `RO_STORAGE`, `TRANSFER_SRC` and `SAMPLED`.
+
+Note that here we split `RO_STORAGE` and `STORAGE` compared to other graphics APIs and that creating a resource with `STORAGE` implies `RO_STORAGE`.
+
+## Render passes
+
+In render passes the only writable resources are textures used as `OUTPUT_ATTACHMENT` and resources used as `STORAGE`.
+
+To avoid data hazards the simplest validation rules would be to check for every subresource that it is used as either:
+ - `OUTPUT_ATTACHMENT`
+ - A combination of read-only usages
+ - A combination of `STORAGE` and `RO_STORAGE`
```

Comment:
in what case does `STORAGE` + `RO_STORAGE` make sense?

Reply:
Using them in two shaders, one that has a read only storage in its bindgroup layout, and the other that has a writable storage in the bindgroup layout.

Reply:
Why would we want to support this though? If one part of a pass writes to a resource and another one reads, then this asks for a transition barrier.